respy/parallelization.py

```diff
@@ -0,0 +1,220 @@
+
+    # Re-order arguments for zipping.
+    args = {idx: [arg[idx] for arg in args] for idx in dense_indices}
+    kwargs = {
+        idx: {kwarg: value[idx] for kwarg, value in kwargs.items()}
+        for idx in dense_indices
+    }
+
+    return args, kwargs
+
+
+def _is_dense_dictionary_argument(argument, dense_indices):
+    """Check whether all keys of the dictionary argument are also dense indices.
+
+    We cannot check whether all dense indices are in the argument because `splitted_df`
+    in :func:`distribute_and_combine_simulation` may not cover all dense combinations.
+
+    """
+    return isinstance(argument, dict) and all(idx in argument for idx in dense_indices)
+
+
+def distribute_and_combine_simulation(func):
+    """Distribute the simulation across sub state spaces and combine."""
+
+    @functools.wraps(func)
+    def wrapper_distribute_and_combine_simulation(df, *args, optim_paras, **kwargs):
+        dense_columns = create_dense_state_space_columns(optim_paras)
+        splitted_df = _split_dataframe(df, dense_columns) if dense_columns else df
+
+        out = func(splitted_df, *args, optim_paras, **kwargs)
+
+        df = pd.concat(out.values()).sort_index() if isinstance(out, dict) else out
+
+        return df
+
+    return wrapper_distribute_and_combine_simulation
+
+
+def distribute_and_combine_likelihood(func):
+    """Distribute the likelihood calculation across sub state spaces and combine."""
+
+    @functools.wraps(func)
+    def wrapper_distribute_and_combine_likelihood(
+        df, base_draws_est, *args, optim_paras, options
+    ):
+        dense_columns = create_dense_state_space_columns(optim_paras)
+        # Duplicate the DataFrame for each type.
+        if dense_columns:
+            n_obs = df.shape[0]
+            n_types = optim_paras["n_types"]
+            # Number each state to split the shocks later. This is necessary to keep the
+            # regression tests from failing.
+            df["__id"] = np.arange(n_obs)
+            # Each row of indices corresponds to a state whereas the columns refer to
+            # different types.
+            indices = np.arange(n_obs * n_types).reshape(n_obs, n_types)
+            splitted_df = {}
+            for i in range(optim_paras["n_types"]):
+                df_ = df.copy().assign(type=i)
+                type_specific_dense = _split_dataframe(df_, dense_columns)
+                splitted_df = {**splitted_df, **type_specific_dense}
+            splitted_shocks = _split_shocks(
+                base_draws_est, splitted_df, indices, optim_paras
+            )
+        else:
+            splitted_df = df
+            splitted_shocks = base_draws_est
+
+        out = func(splitted_df, splitted_shocks, *args, optim_paras, options)
+
+        out = pd.concat(out.values()).sort_index() if isinstance(out, dict) else out
+
+        return out
+
+    return wrapper_distribute_and_combine_likelihood
+
+
+def _split_dataframe(df, group_columns):
```

Comment:
You are completely right! I thought it might be even simpler to loop over the group name and data with
```
for name, group in df.groupby("letter"):
    ...
```